test/server/hot_restarting_base_test.cc

```diff
@@ -0,0 +1,92 @@
+#include "server/hot_restarting_base.h"
+
+#include "test/mocks/api/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/threadsafe_singleton_injector.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Server {
+namespace {
+
+using HotRestartMessage = envoy::HotRestartMessage;
+
+class MockHotRestartingBase : public HotRestartingBase {
+public:
+  MockHotRestartingBase(const std::string& socket_path) : HotRestartingBase(0) {
+    bindDomainSocket(1, "parent", socket_path, 0644);
+  }
+  void sendMessage(sockaddr_un& address, const envoy::HotRestartMessage& message) {
+    sendHotRestartMessage(address, message);
+  }
+};
+
+class HotRestartingBaseTest : public testing::Test {
+public:
+  HotRestartingBaseTest() : base_(path_) {}
+
+  const std::string path_{"/tmp/source"};
+  MockHotRestartingBase base_;
+};
+
+TEST_F(HotRestartingBaseTest, SendMsgRetryFailsAfterRetries) {
+  Api::MockOsSysCalls os_sys_calls;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+
+  uint32_t retries = 0;
+  EXPECT_CALL(os_sys_calls, sendmsg(_, _, _))
+      .WillRepeatedly(Invoke([&](os_fd_t, const msghdr*, int) {
+        ++retries;
+        return Api::SysCallSizeResult{0, ECONNREFUSED};
+      }));
+
+  std::string dst_path = "/tmp/dst";
+  sockaddr_un sun;
+  sun.sun_family = AF_UNIX;
+  StringUtil::strlcpy(&sun.sun_path[1], dst_path.data(), dst_path.size());
+  sun.sun_path[0] = '\0';
+
+  HotRestartMessage message;
+  message.mutable_request()->mutable_pass_listen_socket()->set_address("tcp://0.0.0.0:80");
+
+  EXPECT_DEATH(base_.sendMessage(sun, message), "");
+
+#ifdef NDEBUG
+  // TODO(rgs1): not sure why this isn't visible for release, should be.
+  // EXPECT_EQ(10, retries);
```

Comment:
Hmm. What does this equal in release? Something seems off?

Reply:
Yeah fixing this now, figuring out how to handle side-effects when using EXPECT_DEATH

Reply:
tbh, we don't really need this since the next test properly tests retries. so dropping the retries part, just expecting it to die is good enough.